lib/appmap/event.rb

```diff
@@ -60,16 +60,37 @@ def display_string(value)
           final ? value_string : encode_display_string(value_string)
         end
 
-        def object_properties(hash_like)
-          hash = hash_like.to_h
-          hash.keys.map do |key|
-            {
+        def add_schema(param, value, always: false)
+          return unless always || AppMap.parameter_schema?
+
+          if value.blank? || value.is_a?(String)
+            # pass
+          elsif value.is_a?(Enumerable)
+            if value.is_a?(Hash)
+              param[:properties] = object_properties(value)
+            elsif value.respond_to?(:first) && value.first
+              param[:properties] = object_properties(value.first)
+            end
+          elsif value.respond_to?(:as_json)
+            add_schema param, JSON.parse(value.to_json), always: always
+          end
+        end
+
+        def object_properties(hash)
+          hash = hash.attributes if hash.respond_to?(:attributes)
+          hash = hash.to_h if hash.is_a?(Struct)
+
+          return unless hash.respond_to?(:each_with_object)
+          
+          hash.each_with_object([]) do |entry, memo|
```

Comment:
You could simply do `hash.map { |k, v| { name: k, class: v.class.name } }`. This whole `memo` business is a bit difficult to follow.

Reply:
Unless you do it every day :-)

Reply:
It's very un-rubyish ;)